flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ArrayElementArgumentTypeStrategy.java

```diff
@@ -42,7 +42,7 @@ public Optional<DataType> inferArgumentType(
             CallContext callContext, int argumentPos, boolean throwOnFailure) {
         final ArrayType haystackType =
                 (ArrayType) callContext.getArgumentDataTypes().get(0).getLogicalType();
-        final LogicalType haystackElementType = haystackType.getElementType();
+        final LogicalType haystackElementType = haystackType.getElementType().copy(true);
```

Comment:
it shouldn't be always nullable.
i see 3 cases:
1. `haystackType.getElementType()` is already nullable => no need for force nullable
2. `haystackType.getElementType()` is not nullable and `needleType` is not nullable => no need for force nullable
3.  `haystackType.getElementType()` is not nullable and `needleType` is nullable => need for force nullable.

right now it breaks first 2 cases

Reply:
yeap, you are right. very thanks , and i also add a unit test for the 2nd case, in case of others modify this logical